ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFNullif.java

```diff
@@ -49,6 +50,17 @@ public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumen
     returnOIResolver = new GenericUDFUtils.ReturnObjectInspectorResolver(true);
     returnOIResolver.update(arguments[0]);
 
+    switch (arguments[0].getCategory()) {
+    case LIST:
+    case MAP:
+    case STRUCT:
+    case PRIMITIVE:
+      break;
+    case UNION:
+    default:
+      throw new UDFArgumentTypeException(0, "Unsupported Argument type category: " + arguments[0].getCategory());
+    }
+
     boolean isPrimitive = (arguments[0] instanceof PrimitiveObjectInspector);
     if (isPrimitive)
     {
```

Comment:
When the first argument is a primitive we are making the checks below to ensure various things as the fact they are of exactly the same type. It appears that we are not doing the same for non-primitives, should we?

Reply:
we are only filtering out the `VOID_GROUP` - for complex types I don't think we should do anything beyond checking that the 2 arguments are of the same type

Reply:
From reading the code, it seems that we cannot do `NULLIF(1,'1')` it will fail with an exception. On the contrary it seems that now, we are able to do `NULLIF(array(1,2), array('1','2'))`.

Reply:
OK, my bad was confused with the nesting. Sorry for the confusion